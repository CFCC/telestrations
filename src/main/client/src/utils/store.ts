import { createGamekitSlice, sendEvent } from "@piticent123/gamekit-client";
import {
  configureStore,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";
import {
  TypedUseSelectorHook,
  useSelector as useUntypedSelector,
} from "react-redux";

import { GameState, TelestrationsGame, TelestrationsPlayer } from "./types";
import _ from "lodash";
import { Settings } from "@piticent123/gamekit-client/lib/types";

const { reducer: gamekitReducer, actions: gamekitActions } =
  createGamekitSlice<Settings, TelestrationsPlayer, TelestrationsGame>();

interface AppState {
  gameState: GameState;
  activeContentId: string;
}
interface State {
  app: AppState;
  gamekit: ReturnType<typeof gamekitReducer>;
}
interface ThunkApi {
  state: State;
}

// This has to be at the top level to be throttled correctly
const _sendWritePageEvent = _.throttle((gameCode: string, content: string) => {
  sendEvent({
    route: `/app/games/${gameCode}/page/write`,
    data: content,
  });
}, 500);
export const setGuess = createAsyncThunk<void, string, ThunkApi>(
  "setGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    _sendWritePageEvent(currentGame.code, content);
  }
);

export const submitGuess = createAsyncThunk<void, string, ThunkApi>(
  "submitGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({
      route: `/app/games/${currentGame.code}/page/submit`,
      data: content,
    });
  }
);

export const startGame = createAsyncThunk<void, string, ThunkApi>(
  "startGame",
  async (__, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({ route: `/games/${currentGame.code}/start` });
  }
);

export const {
  createGame,
  joinGame,
  rejoinGame,
  connectToServer,
  saveSettings,
  becomeAdmin,
} = gamekitActions;

const { actions, reducer: appReducer } = createSlice({
  name: "app",
  initialState: {
    gameState: GameState.LOGIN,
    activeContentId: "",
  } as AppState,
  reducers: {
    viewPlayerHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.PLAYER_HISTORY;
      state.activeContentId = action.payload;
    },
    viewNotepadHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.NOTEPAD_HISTORY;
      state.activeContentId = action.payload;
    },
    setGameState: (state, action: PayloadAction<GameState>) => {
      state.gameState = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(gamekitActions.rejoinGame.fulfilled, (state) => {
        state.gameState = GameState.IN_GAME;
      })
      .addCase(gamekitActions.connectToServer.fulfilled, (state) => {
        state.gameState = GameState.GAME_CODE;
      })
      .addCase(gamekitActions.joinGame.fulfilled, (state) => {
        state.gameState = GameState.WAITING_TO_START;
      })
      .addCase(submitGuess.fulfilled, (state) => {
        state.gameState = GameState.WAITING_FOR_CONTENT;
      });
  },
});

export const {
  viewNotepadHistory,
  viewPlayerHistory,
  setGameState
} = actions;

export const store = configureStore({
  reducer: { gamekit: gamekitReducer, app: appReducer },
});

export const useSelector: TypedUseSelectorHook<State> = useUntypedSelector;
