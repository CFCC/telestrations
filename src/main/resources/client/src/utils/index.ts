export function sleep(t: number): Promise<void> {
  return new Promise((res) => setTimeout(res, t));
}
