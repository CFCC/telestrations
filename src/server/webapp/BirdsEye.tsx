import React, {useState, useContext} from "react";
import {
    Card, CardContent, CardHeader,
    Grid,
    Icon,
    IconButton,
    Menu,
    MenuItem,
} from "@material-ui/core";
import PlayerStream from "server/webapp/PlayerStream";
import {UUID} from "types/shared";
import {GameContext} from "server/webapp/Store";
import styled from "styled-components";

interface BirdsEyeState {
    anchorElement: HTMLElement | null;
    playerId: string;
}

const Content = styled(CardContent)`
    padding: 0;
`;

const StyledGrid = styled(Grid)`
    padding: 16px;
`;

export default function BirdsEye() {
    const [{players}, {viewNotepadHistory, viewPlayerHistory}] = useContext(GameContext);
    const [menu, setMenu] = useState({
        anchorElement: null,
        playerId: "",
    } as BirdsEyeState);

    const openMenu = (id: UUID) => (e: React.MouseEvent<HTMLElement, MouseEvent>) =>
        setMenu({anchorElement: e.currentTarget, playerId: id});
    const closeMenu = () => setMenu({anchorElement: null, playerId: ""});

    return (<React.Fragment>
        <StyledGrid container={true} spacing={4}>
            {players.map(player => {
                let playerState: string;
                if (player.notepadIndex === -1) playerState = "Waiting";
                else if (player.notepadIndex % 2 === 1) playerState = "Drawing";
                else /* if (player.notepadIndex % 2 === 0) */ playerState = "Writing";

                return (<Grid item={true} xs={12} sm={6} lg={4} xl={3} key={player.id}>
                    <Card>
                        <CardHeader
                            title={player.nickname}
                            subheader={`Currently ${playerState}`}
                            action={<IconButton
                                onClick={openMenu(player.id)}
                                aria-owns={menu.playerId === player.id ? "menu" : undefined}
                                aria-haspopup="true"
                            >
                                <Icon>more_vert</Icon>
                            </IconButton>}
                        />
                        <Content>
                            <PlayerStream playerId={player.id} />
                        </Content>
                    </Card>
                </Grid>)
            })}
        </StyledGrid>
        <Menu
            open={Boolean(menu.anchorElement)}
            id="menu"
            onClose={closeMenu}
            anchorEl={menu.anchorElement}
        >
            <MenuItem onClick={() => viewPlayerHistory(menu.playerId)}>
                View Player History (Coming Soon!)
            </MenuItem>
            <MenuItem onClick={() => viewNotepadHistory(menu.playerId)}>
                View Notepad History (Coming Soon!)
            </MenuItem>
            <MenuItem>
                Make Stream Fullscreen (Coming Soon!)
            </MenuItem>
        </Menu>
    </React.Fragment>);
}
