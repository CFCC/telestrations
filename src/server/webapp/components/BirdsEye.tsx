import React, {Component} from "react";
import {
    createStyles,
    Grid,
    Icon,
    IconButton,
    Menu,
    MenuItem,
    Paper,
    Typography,
    withStyles,
    WithStyles
} from "@material-ui/core";
import screenfull from "screenfull";
import {State} from "server/webapp/redux/reducers";
import PlayerStream from "server/webapp/components/PlayerStream";
import * as Actions from "server/webapp/redux/actions";
import {ObjectOfRefs} from "types/client";
import {UUID} from "types/shared";
import {connect} from "react-redux";

const styles = createStyles({
    app: {
        padding: "8px"
    }
});

const mapStateToProps = (state: State) => ({
    players: state.players
});

const mapDispatchToProps = {
    viewNotepadHistory: Actions.viewNotepadHistory,
    viewPlayerHistory: Actions.viewPlayerHistory
};

type BirdsEyeProps = WithStyles<typeof styles> & typeof mapDispatchToProps & ReturnType<typeof mapStateToProps>;

interface BirdsEyeState {
    menuAnchorElement: HTMLElement | null;
    menuPlayerId: UUID;
}

@withStyles(styles)
@connect(mapStateToProps, mapDispatchToProps)
export default class BirdsEye extends Component<BirdsEyeProps, BirdsEyeState> {
    state = {
        menuAnchorElement: null,
        menuPlayerId: ""
    };

    streamRefs: ObjectOfRefs = {};

    openMenu = (id: UUID, e: React.MouseEvent<HTMLElement, MouseEvent>) => {
        this.setState({menuAnchorElement: e.currentTarget, menuPlayerId: id});
    };

    closeMenu = () => {
        this.setState({menuAnchorElement: null});
    };

    makeFullScreen = (id: UUID) => {
        if (screenfull.enabled) screenfull.request(this.streamRefs[id]);
    };

    componentDidMount() {
        this.props.players.forEach(p => {
            this.streamRefs[p.id] = React.createRef();
        });
    }

    render() {
        return <div className={this.props.classes.app}>
            <Grid container spacing={8}>
                {
                    this.props.players.map(player => <Grid item xs={12} sm={6} lg={4} xl={3} key={player.id}>
                        <Paper>
                            <Typography variant="h4" gutterBottom>{player.nickname}</Typography>
                            <PlayerStream playerId={player.id} /* ref={this.streamRefs[player.id]} */ />
                            <IconButton onClick={e => this.openMenu(player.id, e)}
                                        aria-owns={this.state.menuPlayerId === player.id ? "menu" : undefined}
                                        aria-haspopup="true">
                                <Icon>more_vert</Icon>
                            </IconButton>
                        </Paper>
                    </Grid>)
                }
            </Grid>
            <Menu open={Boolean(this.state.menuAnchorElement)}
                  id="menu"
                  onClose={this.closeMenu}
                  anchorEl={this.state.menuAnchorElement}>
                <MenuItem onClick={() => this.props.viewPlayerHistory(this.state.menuPlayerId)}>
                    View Player History
                </MenuItem>
                <MenuItem onClick={() => this.props.viewNotepadHistory(this.state.menuPlayerId)}>
                    View Notepad History
                </MenuItem>
                {/*<MenuItem onClick={() => this.makeFullScreen(this.state.menuPlayerId)}>*/}
                {/*    Make Stream Fullscreen*/}
                {/*</MenuItem>*/}
            </Menu>
        </div>;
    }
}
