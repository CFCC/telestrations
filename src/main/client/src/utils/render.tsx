import React from "react";
import { render } from "@testing-library/react";
import { StylesProvider, ThemeProvider } from "@material-ui/core/styles";
import { Provider } from "react-redux";

import { GlobalStyles, theme } from "./theme";
import { store } from "./store";

interface WrapperProps {
  children: React.ReactNode;
}

const wrapper = ({ children }: WrapperProps) => (
  <React.Fragment>
    <StylesProvider injectFirst={true}>
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Provider store={store}>{children}</Provider>
      </ThemeProvider>
    </StylesProvider>
  </React.Fragment>
);
const customRender = (ui: React.ReactElement) =>
  render(ui, { wrapper } as object);

export { customRender as render };
