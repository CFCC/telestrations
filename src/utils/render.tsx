import React from "react";
import {render} from "@testing-library/react";
import {StylesProvider, ThemeProvider} from "@material-ui/core/styles";
import {Provider} from "react-redux";

import {GlobalStyles, theme} from "./theme";
import {store} from "../store/client";

interface WrapperProps {
    children: React.ReactNode;
}

const wrapper = (store: any) => ({children}: WrapperProps) => (
    <React.Fragment>
        <StylesProvider injectFirst={true}>
            <ThemeProvider theme={theme}>
                <GlobalStyles />
                <Provider store={store}>
                    {children}
                </Provider>
            </ThemeProvider>
        </StylesProvider>
    </React.Fragment>
);
const customRender = (ui: React.ReactElement, options: any) => render(ui, {wrapper, ...options});

export {customRender as render};
