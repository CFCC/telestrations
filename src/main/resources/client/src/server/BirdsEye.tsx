import React, { useRef, useState } from "react";
import {
  Card,
  CardContent,
  CardHeader,
  Grid,
  Icon,
  IconButton,
  Menu,
  MenuItem,
  Typography,
} from "@material-ui/core";
import styled from "styled-components";
import { useDispatch } from "react-redux";
import _ from "lodash";

import { actions, useSelector } from "../utils/store";
import PlayerStream from "./PlayerStream";
import * as theme from "../utils/theme";

interface BirdsEyeState {
  anchorElement: HTMLElement | null;
  playerId: string;
}

const Header = styled.div`
  background-color: ${theme.secondary};
  color: white;
  width: calc(100% - 4rem);
  height: 2rem;
  padding: 1rem 2rem;
  display: flex;
  align-items: center;
  justify-content: space-between;
  line-height: 2rem !important;
`;

const Content = styled(CardContent)`
  padding: 0;
`;

const StyledGrid = styled(Grid)`
  padding: 16px;
  width: 100%;
`;

export default function BirdsEye() {
  const gameCode = useSelector((state) => state.currentGame.code);
  const players = useSelector((state) => state.currentGame.players);
  const dispatch = useDispatch();
  const [menu, setMenu] = useState({
    anchorElement: null,
    playerId: "",
  } as BirdsEyeState);
  const fullScreenUser = useRef<HTMLDivElement>(null);

  const closeMenu = () => {
    setMenu({ anchorElement: null, playerId: "" });
  };
  const handleViewPlayerHistory = () => {
    dispatch(actions.viewPlayerHistory(menu.playerId));
  };
  const handleViewNotepadHistory = () => {
    dispatch(
      actions.viewNotepadHistory(
        notepads[players[menu.playerId].currentNotepad].ownerId
      )
    );
  };
  const handleViewFullScreen = () => {
    fullScreenUser.current?.requestFullscreen();
    closeMenu();
  };

  return (
    <React.Fragment>
      <Header>
        <Typography>Game Code: {gameCode}</Typography>
        <Typography variant="h5">Birds Eye View</Typography>
        <Typography>Game {_.startCase(status)}</Typography>
      </Header>
      <StyledGrid container={true} spacing={2}>
        {Object.entries(players).map(([id, player]) => {
          let playerState: string;

          if (!player.currentNotepad) playerState = "Drawing";
          else if (notepads[player.currentNotepad]?.pages?.length % 2 === 1)
            playerState = "Writing";
          else playerState = "Drawing";

          function openMenu(e: React.MouseEvent<HTMLElement, MouseEvent>) {
            setMenu({ anchorElement: e.currentTarget, playerId: id });
          }

          return (
            <Grid item={true} xs={12} sm={6} lg={4} xl={3} key={id}>
              <Card ref={id === menu.playerId ? fullScreenUser : undefined}>
                <CardHeader
                  title={player.name}
                  subheader={`Currently ${playerState}`}
                  action={
                    <IconButton
                      onClick={openMenu}
                      aria-owns={menu.playerId === id ? "menu" : undefined}
                      aria-haspopup="true"
                    >
                      <Icon>more_vert</Icon>
                    </IconButton>
                  }
                />
                <Content>
                  <PlayerStream playerId={id} />
                </Content>
              </Card>
            </Grid>
          );
        })}
      </StyledGrid>
      <Menu
        open={!!menu.anchorElement}
        id="menu"
        onClose={closeMenu}
        anchorEl={menu.anchorElement}
      >
        <MenuItem onClick={handleViewPlayerHistory}>
          View Player History
        </MenuItem>
        <MenuItem onClick={handleViewNotepadHistory}>
          View Notepad History
        </MenuItem>
        <MenuItem onClick={handleViewFullScreen}>
          Make Stream Fullscreen
        </MenuItem>
      </Menu>
    </React.Fragment>
  );
}
