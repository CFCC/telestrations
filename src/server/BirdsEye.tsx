import React, {useState} from "react";
import {
    Card, CardContent, CardHeader,
    Grid,
    Icon,
    IconButton,
    Menu,
    MenuItem,
} from "@material-ui/core";
import styled from "styled-components";

import PlayerStream from "../server/PlayerStream";
import {viewNotepadHistory, viewPlayerHistory, useSelector} from "../store/server";

interface BirdsEyeState {
    anchorElement: HTMLElement | null;
    playerId: string;
}

const Content = styled(CardContent)`
    padding: 0;
`;

const StyledGrid = styled(Grid)`
    padding: 16px;
`;

export default function BirdsEye() {
    const {players, notepads} = useSelector(state => state.game);
    const [menu, setMenu] = useState({
        anchorElement: null,
        playerId: "",
    } as BirdsEyeState);

    const closeMenu = () => setMenu({anchorElement: null, playerId: ""});

    return (
        <React.Fragment>
            <StyledGrid container={true} spacing={4}>
                {Object.entries(players).map(([id, player]) => {
                    let playerState: string;

                    if (!player.currentNotepad)
                        playerState = "Waiting";
                    else if (notepads[player.currentNotepad].pages.length % 2 === 1)
                        playerState = "Writing";
                    else
                        playerState = "Drawing";

                    function openMenu() {
                        return (e: React.MouseEvent<HTMLElement, MouseEvent>) =>
                            setMenu({anchorElement: e.currentTarget, playerId: id})
                    }

                    return (
                        <Grid item={true} xs={12} sm={6} lg={4} xl={3} key={id}>
                            <Card>
                                <CardHeader
                                    title={player.name}
                                    subheader={`Currently ${playerState}`}
                                    action={(
                                        <IconButton
                                            onClick={openMenu}
                                            aria-owns={menu.playerId === id ? "menu" : undefined}
                                            aria-haspopup="true"
                                        >
                                            <Icon>more_vert</Icon>
                                        </IconButton>
                                    )}
                                />
                                <Content>
                                    <PlayerStream playerId={id} />
                                </Content>
                            </Card>
                        </Grid>
                    );
                })}
            </StyledGrid>
            <Menu
                open={Boolean(menu.anchorElement)}
                id="menu"
                onClose={closeMenu}
                anchorEl={menu.anchorElement}
            >
                <MenuItem onClick={() => viewPlayerHistory(menu.playerId)}>
                    View Player History (Coming Soon!)
                </MenuItem>
                <MenuItem onClick={() => viewNotepadHistory(menu.playerId)}>
                    View Notepad History (Coming Soon!)
                </MenuItem>
                <MenuItem>
                    Make Stream Fullscreen (Coming Soon!)
                </MenuItem>
            </Menu>
        </React.Fragment>
    );
}
