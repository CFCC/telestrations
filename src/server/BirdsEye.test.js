import {render} from "../utils/render";
import {Standard} from "./BirdsEye.stories";

it('Renders', () => {
    const {asFragment} = render(Standard());
    expect(asFragment()).toMatchSnapshot();
});
