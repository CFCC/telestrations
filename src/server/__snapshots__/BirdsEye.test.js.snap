// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root sc-AxhUy ljtkiQ MuiGrid-container MuiGrid-spacing-xs-4"
  />
</DocumentFragment>
`;
